---
title: "Setting up a run on mega-non-model"
output: 
  github_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This is a tutorial intended for my esteemed labmates and members of the weekly
MEGA-bioinformatics group.  We are starting to peel a lot of data off of our
own NextSeq machine, and it is important that everyone in the lab get ready
for that firehose.

One of the first orders of business is going to be setting up a simple way to
do some initial/preliminary QA/QC to assess how well these NextSeq runs are
working.  Much of what I will be describing related to this involve things that
I recently added to the mega-non-model workflow to accommodate some of the issues
that we foresee running into.  These include:

- Being able to specify a `data_parent_dir` in the config that will be prefixed
to every path to a fastq file in the units file.  This makes it easier to store
all the necessary files outside of the mega-non-model directory.  (Note that
similar things can be achieved by just using symbolic links within a data
directory, but we also offer the `data_parent_dir` as an option.)
- The addition of a number of new _destination rules_.  These are simply Snakemake
rules that have as input a number of files that one might want to produce for a
specific purpose without necessarily running the whole workflow.  We now have new
destination rules:
  - `dest_prelim_qc`: for preliminary QA/QC that can be done without any read mapping
  (i.e., running fastp and then summarizing the results in various ways).
  - `dest_gvcfs`: for making the gvcfs for every sample, but stopping at that point
  so that other gvcfs can later be made and added to these before loading the genomics
  databases.


We also demonstrate a simple function that we have stored in this repository
(in the `R` directory) that is tailored for converting our NextSeq sample sheets
and sequencing outputs into a `units.tsv` file for running mega-non-model.  

## A note on Snakemake

The mamba package manager has undergone a few breaking changes, so the newest
versions of Snakemake have been using the latest versions of conda (which use
libmamba under the hood now.)  I set that up on one of my laptops, but I am
currently still just using Snakemake 8.5.3 on SEDNA and on my work laptop, and that
seems to be working pretty well with the old version of mamba that I have.

However, we might want to be prepared for glitches for new users.  


## Setting up the units file and config file for mega-non-model

We are going to focus on these two first, because those are the ones
that are needed to do preliminary qc (a chromosomes.tsv and scaffold_groups.tsv
file will need to be there, but won't be relevant.)

### Putting together the units file

This is a critical step, and it is easily done in R from the file listing
of your fastqs.  

First, let's review the mega-non-model instructions regarding this file.
See it [https://github.com/eriqande/mega-non-model-wgs-snakeflow?tab=readme-ov-file#what-the-user-must-do-and-values-to-be-set-etc](here).

To make this file, I got most of the necessary ingredients by using `ls` on SEDNA,
where I have stored the files:
```sh
[sedna: eriq]--% pwd
/share/swfsc/eriq
[sedna: eriq]--% ls -l Landscape_Genomics/*/*.fastq.gz
-rw-rw-r-- 1 eanderson swfsc 1673418258 Mar  2 14:11 Landscape_Genomics/250301_VH02170_2_2227GLJNX/M004147_M028_1A_S62_R1_001.fastq.gz
-rw-rw-r-- 1 eanderson swfsc 1779605042 Mar  2 14:13 Landscape_Genomics/250301_VH02170_2_2227GLJNX/M004147_M028_1A_S62_R2_001.fastq.gz
-rw-rw-r-- 1 eanderson swfsc 1608758499 Mar  2 14:11 Landscape_Genomics/250301_VH02170_2_2227GLJNX/M004182_M028_5D_S61_R1_001.fastq.gz
-rw-rw-r-- 1 eanderson swfsc 1698883931 Mar  2 14:13 Landscape_Genomics/250301_VH02170_2_2227GLJNX/M004182_M028_5D_S61_R2_001.fastq.gz
-rw-rw-r-- 1 eanderson swfsc  845174892 Mar  2 14:11 Landscape_Genomics/250301_VH02170_2_2227GLJNX/M004225_M028_10G_S74_R1_001.fastq.gz
-rw-rw-r-- 1 eanderson swfsc  884389611 Mar  2 14:12 Landscape_Genomics/250301_VH02170_2_2227GLJNX/M004225_M028_10G_S74_R2_001.fastq.gz
-rw-rw-r-- 1 eanderson swfsc  799256079 Mar  2 14:11 Landscape_Genomics/250301_VH02170_2_2227GLJNX/M004226_M028_10H_S75_R1_001.fastq.gz
...

```
and so forth.

I put the results into this tutorial repo at [https://raw.githubusercontent.com/eriqande/mega-bioinf-tutorials/refs/heads/main/data/mega-non/landscape-genomics-fastqs.txt](https://raw.githubusercontent.com/eriqande/mega-bioinf-tutorials/refs/heads/main/data/mega-non/landscape-genomics-fastqs.txt).

Since it is up there, we can actually just use it directly from R, so everyone can follow
along (though you need an internet connection):
```{r, warning=FALSE, message=FALSE}
library(tidyverse)

files_url <- "https://raw.githubusercontent.com/eriqande/mega-bioinf-tutorials/refs/heads/main/data/mega-non/landscape-genomics-fastqs.txt"


files_listing <- read_table(
  files_url,
  col_names = FALSE
)

# we can look at that
files_listing
```

Columns X5 and X9 are what we need.  So, let's rename those and get them:
```{r}
f2 <- files_listing %>%
  rename(fq = X9) %>%
  mutate(kb = as.numeric(X5) / 1024) %>%
  select(fq, kb)

f2
```

Now we want to combine that information with what is in the sample sheet to get columns that
have names like:
```
sample	unit	library	flowcell	platform	lane	sample_id	barcode	fq1	fq2	kb1	kb2
```
Doing this involves just a little rigamoral in R.  I have actually wrapped up all the steps
in some functions that are stored [here]()
